azkaban-common/src/main/java/azkaban/executor/AbstractExecutorManagerAdapter.java

```diff
@@ -599,40 +601,45 @@ protected void handleCancelFlow(ExecutionReference executionReference,
-      ExecutionControllerUtils.finalizeFlow(this, this.projectManager, this.executorLoader,
-          this.alerterHolder,
-          executableFlow, "Cancel action has been called but the flow is unreachable.", null,
+      handleCancelFlowManually(executionReference, executableFlow, finalizingReason, null,
           finalizingStatus);
-      // Killed flow events can only be sent out if callWithReferenceByUser completed successfully
-      // so we need to manually send one here.
-      this.fireEventListeners(Event.create(executableFlow,
-          EventType.FLOW_FINISHED, new EventData(executableFlow)));
-      // Throwing exception to make the reason appear on the UI.
-      throw new ExecutorManagerException("Flow execution is unreachable. Finalizing the flow.");
     }
 
     try {
       this.apiGateway.callWithReferenceByUser(executionReference, ConnectorParams.CANCEL_ACTION,
           userId);
     } catch (Exception e) {
+      final String finalizingReason = "Unable to gracefully kill the flow execution or flow is "
+          + "unreachable.";
       logger
           .error("Exception occurred while cancelling flow: " + executionReference.getExecId(), e);
-      logger.info("Finalizing executable flow: " + executionReference.getExecId());
-      final String finalizingReason = "Unable to gracefully kill the flow execution.";
-      ExecutionControllerUtils.finalizeFlow(this, this.projectManager, this.executorLoader,
-          this.alerterHolder, executableFlow, finalizingReason, e, finalizingStatus);
-      // Killed flow events can only be sent out if callWithReferenceByUser completed successfully
-      // so we need to manually send one here.
-      this.fireEventListeners(Event.create(executableFlow,
-          EventType.FLOW_FINISHED, new EventData(executableFlow)));
-      // Throwing exception to make the reason appear on the UI.
-      throw new ExecutorManagerException(finalizingReason
-          + " Finalizing the flow.");
+      handleCancelFlowManually(executionReference, executableFlow, finalizingReason, e,
+          finalizingStatus);
     }
   }
 
+  private void handleCancelFlowManually(ExecutionReference executionReference,
+      ExecutableFlow executableFlow, String reason, Throwable originalError,
+      Status finalizingStatus) throws ExecutorManagerException {
+    logger.info("Finalizing executable flow: " + executionReference.getExecId());
+    finalizeFlow(executableFlow, reason, originalError, finalizingStatus);
+    // Killed flow events can only be sent out if callWithReferenceByUser completed successfully
+    // so we need to manually send one here.
+    this.fireEventListeners(Event.create(executableFlow,
+        EventType.FLOW_FINISHED, new EventData(executableFlow)));
+    // Throwing exception to make the reason appear on the UI.
+    throw new ExecutorManagerException(reason + " Finalizing the flow.");
+  }
+
+  protected void finalizeFlow(final ExecutableFlow flow, final String reason,
```

Comment:
Same here. I had to look into low level to understand what is finalizeFlow means. it helps if there is a comment saying "mark the flow finish status in DB".